net: Add timeout to various send functions

clamattia opened the pull request:
Add timeout to various send function. To keep stable API, functions are renamed with `_try`-prefix. Original functions just call the `_try`-variant with timeout `K_FOREVER`.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/85981

Timeout can be used to early drop some request for arp and icmp to avoid being slowed when in a flood situation.

Looking forward to your thoughts.

rlubos reviewed (commented):
I think we could go this way if we retain stable APIs compability.

pdgendt commented on subsys/net/ip/net_core.c:
```diff
@@ -363,8 +363,12 @@ static inline int check_ip(struct net_pkt *pkt)
 }
 
-/* Called when data needs to be sent to network */
 int net_send_data(struct net_pkt *pkt)
+{
+	return net_try_send_data(pkt, K_FOREVER);
+}
```
Convert this to an inline function instead?

Commit:
net: add `try` variant with timeout for various send functions

Allows to send with different timeouts to not block caller in some
situations. Stable API is kept and just calls `try`-variant with a timeout
of `K_FOREVER`.

Signed-off-by: Cla Mattia Galliard <cla-mattia.galliard@zuehlke.com>

Commit:
net: net_context: Make use of timeout for sending

Use the newly added timeout in various send functions from within
net_context_sendto.

Signed-off-by: Cla Mattia Galliard <cla-mattia.galliard@zuehlke.com>

Commit:
net: arp: Send replies with timeout `K_NO_WAIT`

This ensures system stays operational in arp-flood situation.

Signed-off-by: Cla Mattia Galliard <cla-mattia.galliard@zuehlke.com>

Commit:
net: lldp: Send replies with timeout `K_NO_WAIT`

This ensures system stays operation in lldp-flood situation.

Signed-off-by: Cla Mattia Galliard <cla-mattia.galliard@zuehlke.com>

Commit:
net: icmp: Send replies with timeout `K_NO_WAIT`

This ensure system stays operational in icmp flood situation.

Signed-off-by: Cla Mattia Galliard <cla-mattia.galliard@zuehlke.com>

clamattia commented:
- Created separate commits for introducing timeout and using it
- made original functions `static inline` as suggested by @pdgendt 


- For now I added timeout for `net_context`, `arp`, `icmp`, `lldp`

clamattia commented on subsys/net/ip/net_context.c:
```diff
@@ -2609,7 +2609,7 @@ static int context_sendto(struct net_context *context,
 						  ntohs(ll_dst_addr->sll_protocol));
 
-			net_if_queue_tx(net_pkt_iface(pkt), pkt);
+			net_if_try_queue_tx(net_pkt_iface(pkt), pkt, timeout);
```
Thanks, will take a look.

clamattia replied:
Added

clamattia commented:
Added timeouts to more send functions in `context_sendto`.

clamattia commented:
Fix compliance issues.

clamattia commented:
- Fix silly mistake
- Document timeout parameter

clamattia commented:
Rebased

clamattia commented:
Reorder declarations in header files, so they are known for static inline functions.

pdgendt reviewed (dismissed):
This looks like a good addition to me.

clamattia commented:
- Fix some CI by using correct `k_timeout_t` type
- Remove no longer timeout to `K_NO_WAIT` in ISR context, because `arp` and `icmp` set to `K_NO_WAIT` in calling context. @rlubos do you agree?

rlubos commented:
> Remove no longer needed timeout to K_NO_WAIT in ISR context, because arp and icmp set to K_NO_WAIT in calling context. @rlubos do you agree?

I'd personally keep it just in case. Just realized, IPv6 ND has probably a similar scenario not handled here (NA reply for NS, in `ipv6_nbr.c`), maybe there are more cases.

clamattia commented:
> > Remove no longer needed timeout to K_NO_WAIT in ISR context, because arp and icmp set to K_NO_WAIT in calling context. @rlubos do you agree?
> 
> I'd personally keep it just in case. Just realized, IPv6 ND has probably a similar scenario not handled here (NA reply for NS, in `ipv6_nbr.c`), maybe there are more cases.

Will add it to the variatn without `try`.
Makes sense. I just implemented for the ones I was familiar with.

clamattia commented:
More compliance/ci fixes

clamattia commented:
Rebased

clamattia commented:
Fixed silly mistake.

clamattia commented:
ci failure likely unrelated. Will try rebasing soon (hopefully fixed in main?).

clamattia commented:
Rebased

clamattia commented:
Is there anything left to do from my side or is this waiting for the feature defrost :) ?

rlubos commented:
> Is there anything left to do from my side or is this waiting for the feature defrost :) ?

AFAICT it's just waiting for the merge window to open.
